Authoring guidelines: description (#171)
codewars/docs: content/recipes/authoring/kata-snippets/description.md

When questions or requests for improvements, clarifications, or examples pop up particularly often, the description should be updated to address them.
- **Descriptions should not make translating a kata into other programming languages difficult.** The description is shared between all translations of the kata and may need to be modified by translators. If there are many translations with many such modifications, description merge conflicts can occur, making kata maintenance difficult. As such, descriptions should ideally be as language-agnostic as possible and use only abstract terms not specific to any programming language. If a description cannot be kept language-agnostic, language-specific elements should be structured in a way which helps with resolution of merge conflicts: preferably in paragraphs or larger blocks rather than multiple language concepts in one sentence.

## Structure

- **Description should be structured into meaningful parts**, for example:

>   - overwiew, general context, or introduction,

Reviewer: ```suggestion
  - overview, general context, or introduction,
```

>   - general statement of a task user has to solve,

Reviewer: ```suggestion
  - general statement of a task the user has to solve,
```

  - description of input and output, potentially with examples,
  - detailed rules and requirements,
  - additional notes, explanations, technical details, follow-ups.  

> Above is just an example; not all sections are necessary if not applicable, and authors can use another structure if they like. However, related pieces of information should be together: explaining inputs in 3-4 different places isn't a good idea.

Reviewer: ```suggestion
The above is just an example; not all sections are necessary unless applicable, and authors can use another structure if they like. However, related pieces of information should be together: explaining inputs in 3-4 different places isn't a good idea.
```

> - **Descriptions should be consise, but precise.** Description of a task should be possibly short, because long descriptions suggest that task is either not very well defined, or is very complex. Fabularized or themed descriptions are OK as long as they don't get in a way of understanding the problem.

Reviewer: ```suggestion
- **Descriptions should be concise, but precise.** The description of a task should be as short as possible, because long descriptions suggest that the task is either not very well defined, or is very complex. Descriptions accompanied with a theme or storyline are acceptable as long as they don't get in the way of understanding the problem.
```

Author: It's a proper word in Polish, English could use it well :)

> - **Descriptions should be consise, but precise.** Description of a task should be possibly short, because long descriptions suggest that task is either not very well defined, or is very complex. Fabularized or themed descriptions are OK as long as they don't get in a way of understanding the problem.

Author: uh, is it even a word, then? :D

> - **Descriptions should be consise, but precise.** Description of a task should be possibly short, because long descriptions suggest that task is either not very well defined, or is very complex. Fabularized or themed descriptions are OK as long as they don't get in a way of understanding the problem.

Reviewer: I learned a new word, "fabular"! :smile: But it seems that "fabularized" is not recognized as a word, at least in some online dictionaries.

I've changed it to simpler English in the meantime.

> - It's usually recommended to describe general aspects first, and then go further into details, edge cases, and finally, exceptions. Descriptions structured this way are usually easier to follow.

Author: No, I meant "inputs which should be handled in some exceptional way", for example,  _"Return index of matching element **or -1 if it could not be found**"_ .

> - It's usually recommended to describe general aspects first, and then go further into details, edge cases, and finally, exceptions. Descriptions structured this way are usually easier to follow.

Reviewer 2: Gotcha, that's what I thought but then "edge cases" maybe covered that--I guess not since edge cases are well-formed but out of the ordinary whereas exceptions would be malformed input of some sort. Maybe "malformed input handling" or something to that effect is a bit clearer?

> - It's usually recommended to describe general aspects first, and then go further into details, edge cases, and finally, exceptions. Descriptions structured this way are usually easier to follow.

Author: Exactly, I avoided "edge cases" because I think they are considered "normal", but on the edge. While malformed inputs could be one case of "exceptional cases" I meant.

Could you please suggest some edit here for this point?

> - It's usually recommended to describe general aspects first, and then go further into details, edge cases, and finally, exceptions. Descriptions structured this way are usually easier to follow.

Reviewer 2: ```suggestion
- It's usually recommended to describe general aspects first, and then go further into details, edge cases and explanations of how malformed input should be handled, if applicable. Descriptions structured this way are usually easier to follow.
```

## Requirements

> - **Description should clearly state all requirements** the user is expected to fulfill to complete the kata. Unless the kata is a puzzle, users should not be given test cases which enforce requirements not mentioned in the description. Description should _explicitly_, _very clearly_ state which inputs have to be handled, which inputs do not have to be considered, and how to handle exceptional situations. When a test case fails, there should always be a point in the description which would explain why the particular test failed.

Reviewer: ```suggestion
- **The description should clearly state all requirements** the user is expected to fulfil to complete the kata. Unless the kata is a puzzle, users should not be given test cases which enforce requirements not mentioned in the description. The description should _explicitly_, _unambiguously_ state which inputs have to be handled, which inputs do not have to be considered, and how to handle exceptional situations. When a test case fails, there should always be a point in the description which would explain why the particular test failed.
```

- Descriptions should provide **examples of inputs and outputs for some representative scenarios**. Provide as many examples as you think necessary.

>   - **Avoid things like _"See sample tests for more examples"_**. To see sample tests, users have to enter the kata trainer, what can mark the kata as unfinished in their profile.

Reviewer: ```suggestion
  - **Avoid things like _"See sample tests for more examples"_**. To see the sample tests, users have to enter the kata trainer, which marks the kata as unfinished in their profile.
```

>   - **Use images when helpful.** Currently, Codewars does not allow to attach files to a kata, so image has to be hosted externally, or HTML image with Base-64 encoded content, preferrably SVG, can be used. Make sure that image does not break layout of the kata page.

Reviewer 2: ```suggestion
  - **Use images when helpful, but prefer text if it's sufficient to communicate the idea.** Currently, Codewars does not allow attaching files to a kata which makes hosting images externally a default solution. Alternatively, a HTML image with Base-64 encoded content or an SVG can be included in the description. This is preferred because external image hosts create a dependency on another website which may become stale over time, rendering the kata unusable. Make sure that image does not break the layout of the kata page. When using images, take into consideration users with small screen widths and visual impairments.
```

>   - **Use images when helpful, but prefer text if it's sufficient to communicate the idea.** Currently, Codewars does not allow attaching files to a kata which makes hosting images externally a default solution. Alternatively, a HTML image with Base-64 encoded content or an SVG can be included in the description. This is preferred because external image hosts create a dependency on another website which may become stale over time, rendering the kata unusable. Make sure that image does not break the layout of the kata page. When using images, take into consideration users with small screen widths and visual impairments.

Reviewer: ```suggestion
  - **Use images when helpful, but prefer text if it's sufficient to communicate the idea.** Currently, Codewars does not allow attaching files to a kata which makes hosting images externally a default solution. Alternatively, a HTML image with Base-64 encoded content or an SVG can be included in the description. This is preferred because external image hosts create a dependency on another website which may become stale over time, rendering the kata unusable. Make sure that the image does not break the layout of the kata page. When using images, take into consideration users with small screen widths and visual impairments.
```

> - **Describe or hint the difficulty.** It's a very frustrating experience when a user comes up with solution returning correct answers, only to find out that it times out upon submission. Provide information about input ranges, volume of tests, expected complexity class, etc. 

Reviewer: ```suggestion
- **Describe or hint on the difficulty.** It's a very frustrating experience when a user comes up with a solution returning the correct answers, only to find out that it times out upon submission. Provide information about input ranges, volume of tests, expected complexity class, etc.
```

Not 100% sure about adding the "on" there ...

## Formatting

> - **Use [Markdown capabilities](/references/markdown/)** to format and structure descriptions appropriately. You can use headers and lines to introduce sections, lists and tables to collect various items, links to provide references, code markup to introduce code inline or in blocks, and basic text formatting for styling. Do not abuse it though, avoid visual noise and excessive formatting.

Reviewer 2: ```suggestion
- **Use [Markdown capabilities](/references/markdown/)** to format and structure descriptions appropriately. You can use headers and lines to introduce sections, lists and tables to collect various items, links to provide references, code markup to introduce code inline or in blocks, and basic text formatting for styling. Do not abuse it though; avoid visual noise and excessive formatting.
```

- **Use proper header levels.** Do not use `# Title` to introduce sections, consider `## Second level` or `### Third level` headers for organization.
- **Use [Sequential Code Blocks](/references/markdown/extensions/#sequential-code-blocks) for blocks of code, examples, etc.

> - **Avoid descriptions mentioning many languages.** Use [Language Conditional Rendering](/references/markdown/extensions/#conditional-rendering) for language specific paragraphs and sections. Do not stuff many languages into a single sentence or paragraph (_"Return `null` for Java/C#/Scala, `None` for Python, `Nothing` for Haskell, `NULL` for C/C++, `nil` for Swift/Ruby"_).

Reviewer 2: ```suggestion
- **Avoid descriptions mentioning many languages.** Use [Language Conditional Rendering](/references/markdown/extensions/#conditional-rendering) for language-specific paragraphs and sections. Including multiple languages in a single sentence or paragraph (_"Return `null` for Java/C#/Scala, `None` for Python, `Nothing` for Haskell, `NULL` for C/C++, `nil` for Swift/Ruby"_) is noisy and harms description readability.
```

> - **Avoid HTML.** Interactions between embedded HTML and Markdown or theming are not well defined and can break in unexpected way. HTML used for styling can give wrong results after CSS is updated or can not work well with visual themes (for example dark or light mode). HTML **can** be used as long as there's a valid purpose and the author is aware of the downsides.

Reviewer 2: ```suggestion
- **Avoid HTML.** Interactions between embedded HTML and Markdown or theming are not well-defined and can break unexpectedly. HTML used for styling can give wrong results after CSS is updated or can not work well with visual themes (for example dark or light mode). HTML **can** be used as long as there's a valid purpose and the author is aware of the downsides.
```

## Other

> - **Description can contain links and references** to external sources of knowledge, articles, honts, explanations, papers.

Reviewer 2: ```suggestion
- **Description can contain links and references** to external sources of knowledge, articles, explanations and papers. Keep in mind that external links create dependencies. Links that change over time may render a kata unusable and introduce a maintenance issue. 
```

Reviewer 2: I'm not sure if "honts" was intended to be "fonts".

Author: It was meant to be "hints" :)

Reviewer 2: That makes more sense than fonts!

> - **Description can contain links to related kata**: other variants of the task, difficulty levels, challenge versions. Links should go in both directions, for example easier kata should link to difficult one, and vice versa.

Reviewer 2: ```suggestion
- **Description can contain links to related kata**: other variants of the task, difficulty levels and challenge versions.
```

Author: I wanted to be explicit about links going in both directions, because it sometimes happened that I wanted to compare two versions of a problem (easier and harder), but the one I was looking at was missing the link.

Reviewer 2: Sure, feel free to keep it in.

> - **Consider adding necessary preloaded code to the description.** As explained in a (related part)(/recipes/authoring/kata-snippets/preloaded/#accessibility-of-preloaded-code), preloaded code is not visible to users attempting the kata, so they cannot easily set up their local environment to work on the attempted kata. It's strongly recommended to help them out by providing them with all the missing pieces needed to solve the kata in local IDE.

Reviewer 2: ```suggestion
- **Consider adding necessary preloaded code to the description.** As explained in a (related part)(/recipes/authoring/kata-snippets/preloaded/#accessibility-of-preloaded-code), preloaded code is not visible to users attempting the kata, so they cannot easily set up their local environment to work on the attempted kata. It's strongly recommended to provide everything necessary to solve the kata in a local IDE.
```

> - **Description can contain links to related kata**: other variants of the task, difficulty levels, challenge versions. Links should go in both directions, for example easier kata should link to difficult one, and vice versa.

Reviewer: ```suggestion
- **Description can contain links to related kata**: other variants of the task, difficulty levels, challenge versions. Links should go in both directions, for example the easier kata should link to the more difficult version and vice versa.
```

> - **Consider adding necessary preloaded code to the description.** As explained in a (related part)(/recipes/authoring/kata-snippets/preloaded/#accessibility-of-preloaded-code), preloaded code is not visible to users attempting the kata, so they cannot easily set up their local environment to work on the attempted kata. It's strongly recommended to provide everything necessary to solve the kata in a local IDE.

Reviewer 3: ```suggestion
- **Consider adding necessary preloaded code to the description.** As explained in a [related part](/recipes/authoring/kata-snippets/preloaded/#accessibility-of-preloaded-code), preloaded code is not visible to users attempting the kata, so they cannot easily set up their local environment to work on the attempted kata. It's strongly recommended to provide everything necessary to solve the kata in a local IDE.
```
